### PR TITLE
Worker propagates errors when writing to disk

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes https://github.com/dask/distributed/issues/1088

### CODE REVIEW

File: distributed/tests/test_worker.py [hunk 1]
```diff
@@ -736,3 +736,21 @@ def test_worker_dir(c, s, a, b):
...
+        def __sizeof__(self):
+            return int(100e9)
+
+    future = c.submit(Bad)
+    yield _wait(future)
+
+    assert future.status == 'error'
+
+    with pytest.raises(TypeError):
+        yield future._result()
```
Comment (pitrou) [inline distributed/tests/test_worker.py]: Perhaps test that we can submit another task after this one?

Comment (mrocklin) [inline distributed/tests/test_worker.py]: Done

**Comment by mrocklin** [comment]
@rabernat this may affect you

File: distributed/worker.py [hunk 2]
```diff
@@ -1365,8 +1366,17 @@ def transition_executing_done(self, key, value=no_value):
...
+                    self.put_key_in_memory(key, value, transition=False)
+                except Exception as e:
+                    logger.info("Failed to put key in memory", exc_info=True)
+                    msg = error_message(e)
+                    self.exceptions[key] = msg['exception']
+                    self.tracebacks[key] = msg['traceback']
+                    self.task_state[key] = 'error'
+                    out = 'error'
+
                 if key in self.dep_state:
```
Comment (pitrou) [inline distributed/worker.py]: Should this still be executed on error?

**Comment by mrocklin** [comment]
OK, I've added a test that fails in a way that I'm not sure how best to fix.

When a piece of data fits in memory, and then later gets pushed to disk what should we do?  Should we keep it in memory?  Should we change the state from `memory` to `error`?

**Comment by pitrou** [comment]
If we wanted to push it out from memory, then my intuition is that it should end up in the `error` state.

**Comment by mrocklin** [comment]
I find myself a little bit stalled here.  Adding a transition from memory->error on the scheduler side is somewhat expensive for such a rarely occurring case.  It's hard to justify.  At the same time it's not clear that a very clean solution exists otherwise.  Still thinking about this.

**Comment by pitrou** [comment]
>  Adding a transition from memory->error on the scheduler side is somewhat expensive for such a rarely occurring case.

Expensive in what terms? If the case is rarely occurring, do we mind the expense?

**Comment by mrocklin** [comment]
Expensive in terms of development and maintenance.

**Comment by mrocklin** [comment]
OK, I've handled this by generalizing a bit.  

The workers already had a system that would flag when they couldn't obtain a dependency.  Previously they would kick this back up to the scheduler giving up and ask the scheduler to recompute elsewhere.  The original need for this behavior was removed a while ago (handled through other means).  

Now instead of asking the scheduler to compute elsewhere they say that the task has erred.  This covers our current case (and others).  I don't think that it will affect any existing situation, but we should be aware of this change if bugs come in soon afterwards.  Stress tests pass.

**Comment by mrocklin** [comment]
Merging this soon if there are no further comments.

**Pull request merged**